### PR TITLE
Switch to async/await (fixes #538)

### LINKED ISSUE #538: Switch to async/await
Revamp PR #45 now that babelisation is more stable :)

### PR DESCRIPTION
**Pull request by leplatrem**

### CODE REVIEW

**Comment by leplatrem** [comment]
@n1k0 the migration was smooth and easy.

There are 4 fours remaining tests that fail, the problem seems to be very similar (error not thrown)... Could you please give me a hand?

**Comment by n1k0** [comment]
I'm looking at it!

File: test/collection_test.js [hunk 1]
```diff
@@ -2171,8 +2175,8 @@ describe("Collection", () => {
     });
 
     it("should validate the remote option", () => {
-      expect(() => articles.sync({remote: "http://fake.invalid"}))
-        .to.Throw(Error, /contain the version/);
+      return articles.sync({remote: "http://fake.invalid"})
+        .should.be.rejectedWith(Error, /contain the version/);
```
Comment (n1k0) [inline test/collection_test.js]: So this means we're actually breaking the public API; previously we were synchronously throwing, but now that `sync()` is declared entirely async, it's not possible anymore. We should add a note about this in the release changelog.

**Comment by n1k0** [comment]
The diff is a pain to review because indentation changes make large chunks of red and green. I'm just gonna proceed with trust and give this an r+ 👍

**Comment by glasserc** [comment]
You can add a `?w=1` argument to diffs on Github to diff with whitespace-insensitivity. That might be useful on this PR.

File: src/collection.js [hunk 2]
```diff
@@ -852,120 +840,104 @@ export default class Collection {
...
+        if (r._status === "created") {
+          batch.createRecord(published);
+        } else {
+          batch.updateRecord(published);
         }
       });
+    }, {headers: options.headers, safe, aggregate: true});
+
+    // Store outgoing errors into sync result object
+    syncResultObject.add("errors", synced.errors.map(e => ({...e, type: "outgoing"})));
```
Comment (glasserc) [inline src/collection.js]: Nice immutability cleanup.

File: src/collection.js [hunk 3]
```diff
@@ -852,120 +840,104 @@ export default class Collection {
...
+    syncResultObject.add("conflicts", conflicts);
+
+    // Reflect publication results locally using the response from
+    // the batch request.
+    // For created and updated records, the last_modified coming from server
+    // will be stored locally.
+    // Records that must be deleted are either deletions that were pushed
+    // to server (published) or deleted records that were never pushed (skipped).
+    const missingRemotely = synced.skipped.map(r => ({...r, deleted: true}));
+    const published = synced.published.map((c) => c.data);
```
Comment (glasserc) [inline src/collection.js]: I think this belongs with the "reflect publication results locally" comment.

**Pull request merged**